packages/core/epoch/types.go

```diff
@@ -167,3 +172,129 @@ func (e *ECRecord) SetPrevEC(prevEC EC) {
 	e.M.PrevEC = NewMerkleRoot(prevEC[:])
 	e.SetModified()
 }
+
+type NodesActivityLog map[identity.ID]*ActivityLog
+
+// region ActivityLog //////////////////////////////////////////////////////////////////////////////////////////////////
+
+// ActivityLog is a time-based log of node activity. It stores information when a node was active and provides
+// functionality to query for certain timeframes.
+type ActivityLog struct {
+	SetEpochs set.Set[Index] `serix:"0,lengthPrefixType=uint32"`
+}
+
+// NewActivityLog is the constructor for ActivityLog.
+func NewActivityLog() *ActivityLog {
+
+	a := &ActivityLog{
+		SetEpochs: set.New[Index](),
+	}
+
+	return a
+}
+
+// Add adds a node activity to the log.
+func (a *ActivityLog) Add(ei Index) (added bool) {
+	if !a.SetEpochs.Add(ei) {
+		return false
+	}
+
+	return true
+}
+
+// Remove removes a node activity from the log.
+func (a *ActivityLog) Remove(ei Index) (removed bool) {
+	if !a.SetEpochs.Delete(ei) {
+		return false
+	}
+
+	return true
+}
+
+// Active returns true if the node was active between lower and upper bound.
+// It cleans up the log on the fly, meaning that old/stale times are deleted.
+// If the log ends up empty after cleaning up, empty is set to true.
+func (a *ActivityLog) Active(lowerBound, upperBound Index) (active bool) {
+	for ei := lowerBound; ei <= upperBound; ei++ {
+		if a.SetEpochs.Has(ei) {
+			active = true
+		}
+	}
```

Comment:
When is the clean-up happening? 
Return as soon as there's an active:
```go
if a.SetEpochs.Has(ei) {
	return true
}
```

Reply:
It is cleaned afterward in a separate method `Clean` I forgot to update the comment